### DISCUSSION TITLE
Will there be a REST API for the FlowForge platform?

### DISCUSSION THREAD

**Discussion by Sulomus** [FlowFuse/flowfuse#1575] [Ideas]
Hello,
we have a customers that want to monitor 10s to 100s of devices.

- We would like to use FlowForge for such usecases. The problem is that we would like to automate the process of installing FlowForge agent, linking it to FlowForge cloud, and registering the device.
- The main building block to this process would be a `REST API` through which we could do basic create, read, update operations on the devices.

Will this be supported in the future?

**Comment by hardillb** [comment]
We are looking at options for bulk enrolling/configuring devices.

First pass may be uploading something like a CSV file with device details and getting back a bundle of device credential files. But exposing the existing API is also possible (we would just need a way to issue API tokens)

**Reply by Sulomus** [reply]
The CSV file would be nice to start with! 😄 
But I think the API will give you more _granulin_ way of handling some operations.

I was thinking that with REST API you can create a script to which you pass the admin token, and it will:

1. install the FlowForge agent onto your edge device.
2. create the device.yaml file and uploads it into the edge device.
3. and connects the edge device to FlowForge.

Basically it would do all the manual work for you (something that the uploading of CSV file couldn't).

**Comment by knolleary** [comment]
Hi @Sulomus 

We do have a REST API for the platform, but it is not currently documented, so not the easiest thing for 3rd parties to work with right now.

I think there are two issues you could follow on this:

1. #1556 - this is to create Swagger/OpenAPI docs for the API. That will make it much easier to consume.
2. #14 - this is to add the ability to create API access tokens that can be used when scripting access to the API. You can work around this today by extracting the session token for a regular login session... but that isn't ideal as the session tokens expire.


On your specific use case, we also have an item on the backlog to provide a way to auto-enroll devices - #1212. Would be interested on your thoughts on the proposed approach in that issue.
